web/src/navigator.rs

```diff
@@ -404,42 +405,55 @@ impl NavigatorBackend for WebNavigatorBackend {
             }
         };
 
-        let (mut sink, mut stream) = ws.split();
+        let (mut ws_write, mut ws_read) = ws.split();
         sender
             .try_send(SocketAction::Connect(handle, ConnectionState::Connected))
             .expect("working channel send");
 
-        // Spawn future to handle incoming messages.
-        let stream_sender = sender.clone();
         self.spawn_future(Box::pin(async move {
-            while let Some(msg) = stream.next().await {
-                match msg {
-                    Ok(Message::Bytes(buf)) => stream_sender
-                        .try_send(SocketAction::Data(handle, buf))
-                        .expect("working channel send"),
-                    Ok(_) => tracing::warn!("Server sent unexpected text message"),
-                    Err(_) => {
-                        stream_sender
-                            .try_send(SocketAction::Close(handle))
-                            .expect("working channel send");
-                        return Ok(());
+            loop {
+                match future::select(ws_read.next(), std::pin::pin!(receiver.recv())).await {
+                    // Handle incoming messages.
+                    Either::Left((msg, _)) => {
+                        let Some(msg) = msg else {
```

Comment:
I think you can probably match on `Some(msg)` specifically (here and below for outgoing), and then`_` to `break`?

Reply:
that's right, looks better now :)